From 557cff85b9cdebb9151aad5c37aff30f3868cf13 Mon Sep 17 00:00:00 2001
From: Konstantin Porotchkin <kostap@marvell.com>
Date: Thu, 24 May 2018 13:39:46 +0300
Subject: [PATCH 0627/1051] mmc: xenon: Add support for AP806-B0 SDHCI
 interface

The AP806 starting from revision B0 supports full HS400 speed
mode on its SDIO interface over 8-bit bus.
The Xenon driver will continue to limit SDIO speed on older
releases of AP806.
The Xenon PHY driver will also activate the slow mode on
AP806 release prior to B0.

Change-Id: Ibbd732ebc2fe9a4d13ba08d0b3218c30fa4d300b
Signed-off-by: Konstantin Porotchkin <kostap@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/mmc/host/sdhci-xenon-phy.c | 8 +++++++-
 drivers/mmc/host/sdhci-xenon.c     | 8 ++++++--
 2 files changed, 13 insertions(+), 3 deletions(-)

diff --git a/drivers/mmc/host/sdhci-xenon-phy.c b/drivers/mmc/host/sdhci-xenon-phy.c
index caccedc836dc..4e228227bd09 100644
--- a/drivers/mmc/host/sdhci-xenon-phy.c
+++ b/drivers/mmc/host/sdhci-xenon-phy.c
@@ -14,6 +14,7 @@
 #include <linux/slab.h>
 #include <linux/delay.h>
 #include <linux/ktime.h>
+#include <linux/mv_soc_info.h>
 #include <linux/of_address.h>
 
 #include "sdhci-pltfm.h"
@@ -699,7 +700,12 @@ static int xenon_emmc_phy_parse_param_dt(struct sdhci_host *host,
 	u32 value;
 
 	params->slow_mode = false;
-	if (of_property_read_bool(np, "marvell,xenon-phy-slow-mode"))
+	/* Activate slow mode unconditionally for AP806 A0 and A1, but not B0.
+	 * For any revision honor the explicit request for the slow mode
+	 */
+	if ((of_device_is_compatible(np, "marvell,armada-ap806-sdhci") &&
+	    mv_soc_info_get_ap_revision() < APN806_REV_ID_B0) ||
+	    of_property_read_bool(np, "marvell,xenon-phy-slow-mode"))
 		params->slow_mode = true;
 
 	params->znr = XENON_ZNR_DEF_VALUE;
diff --git a/drivers/mmc/host/sdhci-xenon.c b/drivers/mmc/host/sdhci-xenon.c
index a0b5089b3274..14e8891dacc7 100644
--- a/drivers/mmc/host/sdhci-xenon.c
+++ b/drivers/mmc/host/sdhci-xenon.c
@@ -17,6 +17,7 @@
 #include <linux/delay.h>
 #include <linux/ktime.h>
 #include <linux/module.h>
+#include <linux/mv_soc_info.h>
 #include <linux/of.h>
 #include <linux/pm.h>
 #include <linux/pm_runtime.h>
@@ -410,9 +411,12 @@ static int xenon_probe_dt(struct platform_device *pdev)
 	u32 sdhc_id, nr_sdhc;
 	u32 tuning_count;
 
-	/* Disable HS200 on Armada AP806 */
-	if (of_device_is_compatible(np, "marvell,armada-ap806-sdhci"))
+	/* Disable HS200 on Armada AP806 A0 and A1, but not B0 */
+	if (of_device_is_compatible(np, "marvell,armada-ap806-sdhci") &&
+	    mv_soc_info_get_ap_revision() < APN806_REV_ID_B0) {
+		dev_err(mmc_dev(mmc), "AP SDHC is running in slow mode\n");
 		host->quirks2 |= SDHCI_QUIRK2_BROKEN_HS200;
+	}
 
 	sdhc_id = 0x0;
 	if (!of_property_read_u32(np, "marvell,xenon-sdhc-id", &sdhc_id)) {
-- 
2.17.1

